src/stream/src/executor/subscription.rs

```diff
@@ -0,0 +1,196 @@
+                        while let Some(row) = data_iter.next().await {
+                            let row = row?;
+                            let now_epoch = row
+                                .datum_at(1)
+                                .ok_or_else(|| {
+                                    StreamExecutorError::from(
+                                        "Can't find epoch from subscription".to_string(),
+                                    )
+                                })?
+                                .into_decimal();
+                            epoch = epoch.max(
+                                u64::try_from(now_epoch)
+                                    .map_err(|err| StreamExecutorError::from(err.to_string()))?,
+                            );
+                            delete_row.push(row);
+                        }
+                    }
+                    delete_row.into_iter().for_each(|row| {
+                        self.epoch_store.delete(row.as_ref());
+                    });
+                    self.log_store.truncate(TruncateOffset::Barrier { epoch })?;
+                    continue;
+                }
+                Either::Right(msg) => {
+                    let msg = msg?;
+                    match msg {
+                        Message::Watermark(w) => Message::Watermark(w),
+                        Message::Chunk(chunk) => {
+                            if chunk.cardinality() == 0 {
+                                // empty chunk
+                                continue;
+                            }
+                            self.log_store.write_chunk(chunk.clone())?;
+                            Message::Chunk(chunk)
+                        }
+                        Message::Barrier(barrier) => {
+                            self.log_store
+                                .flush_current_epoch(
+                                    barrier.epoch.curr,
+                                    barrier.kind.is_checkpoint(),
+                                )
+                                .await?;
+
+                            if let Some(vnode_bitmap) =
+                                barrier.as_update_vnode_bitmap(self.actor_context.id)
+                            {
+                                self.log_store.update_vnode_bitmap(vnode_bitmap)?;
+                            }
+                            let row = OwnedRow::new(vec![
+                                Some(ScalarImpl::Timestamptz(Timestamptz::from_now())),
+                                Some(ScalarImpl::Decimal(barrier.epoch.curr.into())),
+                            ]);
+                            self.epoch_store.insert(row);
+                            self.epoch_store.commit(barrier.epoch).await?;
+                            Message::Barrier(barrier)
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    #[try_stream(ok = (), error = StreamExecutorError)]
```

Comment:
I think this side of timeout stream is unnecessary, because even if the upstream has no data, barrier will always come.